app/com/thumbtack/becquerel/datasources/elasticsearch/EsRetryHttpClient.scala

```diff
@@ -87,12 +101,18 @@ class EsRetryHttpClient(
       logger.warn(s"Retrying rate-limited Elasticsearch request in $wait (attempt $attempts of $maxAttempts)")
 
       val promise = Promise[U]()
-      scheduler.scheduleOnce(wait) {
+      val task: Cancellable = scheduler.scheduleOnce(wait) {
         promise.completeWith(
           safeExecute(request)
             .recoverWith(retry(request, attempts + 1))
         )
       }
+      pendingTasks.synchronized {
+        pendingTasks += task
+      }
+      promise.future.onComplete(_ => pendingTasks.synchronized {
```

Comment:
if `task` completes before this line executed, will this code execute correctly? (e.g. `pendingTasks -= task` should be done anyway)

Reply:
Yes. See the `Future.onComplete` docs:

    If the future has already been completed, this will either be applied immediately or be scheduled asynchronously.